### ISSUE TITLE
Override gorm tags

### ISSUE THREAD

**Issue by hervit0** [t-tiger/gorm-bulk-insert#29]
Hi there,

Thanks a lot for putting together a bulk insert for gorm.

I would like to highlight an issue we experienced, let's say we have a struct with a field defined such as:
```
type fakeTable struct {
	TossACoin string `gorm:"column:toYourWitcher"`
}
```

Because there is this [line](https://github.com/t-tiger/gorm-bulk-insert/blob/0f3bfc1033efe45dd6cd07792c8cfcff48f30d73/bulk_insert.go#L49):
```
dbColumns = append(dbColumns, mainScope.Quote(gorm.ToColumnName(key)))
```

The insert statement will be 
```
...  INSERT INTO table (`toss_a_coin`,  ... 
```
which is the default naming strategy of gorm.

The expected statement will be:
```
...  INSERT INTO table (`toYourWitcher`,  ... 
```

Please, is there a way of not overriding the column names provided by tags when present?

**Comment by bombsimon** [comment]
Thanks for this! This looks like a bug, I'll try to look into this asap.

**Comment by bombsimon** [comment]
I think I might have misunderstood this question. I thought the problem here was that with the above struct tag you were getting `INSERT INTO table ('toss_a_coin',  ... ` but that's not the case for me.

If I apply this change:

```diff
diff --git a/bulk_insert_test.go b/bulk_insert_test.go
index a1a77fc..1878036 100644
--- a/bulk_insert_test.go
+++ b/bulk_insert_test.go
@@ -16,7 +16,7 @@ type fakeRelationTable struct{}
 type fakeTable struct {
        ID        int `gorm:"primary_key;auto_increment"`
        Name      string
-       Email     string              `gorm:"default:default@mail.com"`
+       Email     string              `gorm:"column:this_is_custom; default:default@mail.com"`
        Relation  *fakeRelationTable  `gorm:"foreignkey:RelationID"`
        Relations []fakeRelationTable `gorm:"foreignkey:UserRefer"`
        Message   sql.NullString
```

The test will fail because the column name is bad:

```text
=== RUN   Test_extractMapValue
    Test_extractMapValue: bulk_insert_test.go:67:
                Error Trace:    bulk_insert_test.go:67
                Error:          Not equal:
                                expected: []string{"created_at", "email", "message", "name", "publish", "updated_at"}
                                actual  : []string{"created_at", "message", "name", "publish", "this_is_custom", "updated_at"}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -2,3 +2,2 @@
                                  (string) (len=10) "created_at",
                                - (string) (len=5) "email",
                                  (string) (len=7) "message",
                                @@ -6,2 +5,3 @@
                                  (string) (len=7) "publish",
                                + (string) (len=14) "this_is_custom",
                                  (string) (len=10) "updated_at"
                Test:           Test_extractMapValue
```

And if I print the SQL it reads:

```text
INSERT INTO `fake_tables` (`created_at`, `message`, `name`, `publish`, `this_is_custom`, `updated_at`) VALUES (?, ?, ?, ?, ?, ?)
```

So I don't think I'm sure exactly what you mean. Do you expect the column to be named `email` even though the tag says `i_am_custom`? Or do you see `email` even with the tag applied?

Please provide some more information.

**Comment by hervit0** [comment]
Hello @bombsimon 

Thanks a ton for having a look to this case.

To clarify:
- if I have this tag defined in the struct `Email string gorm:"column:thisIsCustom`
- I expect the query to be:
```
INSERT INTO `fake_tables` (`thisIsCustom`, ... )
```

The test you've modified is giving you the desired outcome because it's testing the `extractMapValue` method in isolation, I think.

However, I think the culprit is `mainScope.Quote(gorm.ToColumnName(key))`, if you provide a custom camel case name, it will default it in underscore case, which is the issue we experienced.

Let me know if you need more details, sorry that I wasn't clear, I will amend the issue description.

**Comment by bombsimon** [comment]
@hervit0 Ah, gotcha!

This is a part of `gorm` which applies with or without this wrapper. You should have a look at the [Namer](https://pkg.go.dev/github.com/jinzhu/gorm?tab=doc#Namer) signature and apply that to [TheNamingStrategy](https://pkg.go.dev/github.com/jinzhu/gorm?tab=doc#NamingStrategy).

```go
gorm.TheNamingStrategy.Column = func(original string) string {
    // Optinally add conditions here, `if original == "thisIsCustom"`
    return original
}

db, _ := gorm.Open("mysql", "")
gormbulk.insertObjSet(db, []interface{}{value})
```

So instead given a camel case name:

```diff
iff --git a/bulk_insert_test.go b/bulk_insert_test.go
index a1a77fc..bb95dee 100644
--- a/bulk_insert_test.go
+++ b/bulk_insert_test.go
@@ -16,7 +16,7 @@ type fakeRelationTable struct{}
 type fakeTable struct {
        ID        int `gorm:"primary_key;auto_increment"`
        Name      string
-       Email     string              `gorm:"default:default@mail.com"`
+       Email     string              `gorm:"column:camelCaseForAll; default:default@mail.com"`
        Relation  *fakeRelationTable  `gorm:"foreignkey:RelationID"`
        Relations []fakeRelationTable `gorm:"foreignkey:UserRefer"`
        Message   sql.NullString
```

We now get 

```text
--- FAIL: Test_extractMapValue (0.00s)
    bulk_insert_test.go:67:
                Error Trace:    bulk_insert_test.go:67
                Error:          Not equal:
                                expected: []string{"created_at", "email", "message", "name", "publish", "updated_at"}
                                actual  : []string{"camelCaseForAll", "created_at", "message", "name", "publish", "updated_at"}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,4 +1,4 @@
                                 ([]string) (len=6) {
                                + (string) (len=15) "camelCaseForAll",
                                  (string) (len=10) "created_at",
                                - (string) (len=5) "email",
                                  (string) (len=7) "message",
                Test:           Test_extractMapValue
```

And the SQL

```sql
INSERT INTO `fake_tables` (`camelCaseForAll`, `created_at`, `message`, `name`, `publish`, `updated_at`) VALUES (?, ?, ?, ?, ?, ?)
```

So just set this up when importing `gorm` and creating the connection you'll use with `gormbulk` and this should hopefully resolve your issue.

**Comment by bombsimon** [comment]
@hervit0 I've created a PR that shows how to keep the custom column name and also ensures that no future changes will break this support/feature. It may also serve as an example for your use case even though it's more or less the same as my post above.

**Comment by bombsimon** [comment]
Maybe I was too quick with this one, I didn’t test if native gorm uses the tag as is. 

Sorry for the confusion, I’ll try to see how native gorm is handling tags tomorrow, you can se the code above as a workaround in the meantime.

**Comment by bombsimon** [comment]
Yeah I was too quick, native gorm uses the column name as is so I'll have another look at this issue.

<details><summary>Code to prove this</summary>

```go
package main

import (
	"testing"

	"github.com/DATA-DOG/go-sqlmock"
	"github.com/jinzhu/gorm"
	"github.com/stretchr/testify/require"
	gormbulk "github.com/t-tiger/gorm-bulk-insert"
)

type Table struct {
	RegularColumn string
	Custom        string `gorm:"column:ThisIsCamelCase"`
}

func Test_insertObject(t *testing.T) {
	// Setup mock database and a gorm instance with said connection.
	db, mock, err := sqlmock.New()
	require.NoError(t, err)

	defer db.Close()

	gdb, err := gorm.Open("mysql", db)
	require.NoError(t, err)

	gdb.Debug()

	// Save the original naming strategy for the columns while we test.
	defaultColumnStrategy := gorm.TheNamingStrategy.Column

	// Set a custom naming strategy which will preserv a given column name,
	// otherwise the default strategy will be used.
	gorm.TheNamingStrategy.Column = func(original string) string {
		if original == "ThisIsCamelCase" {
			return original
		}

		return defaultColumnStrategy(original)
	}

	// With gormbulk we expect the statement to be sorted in order and that the
	// given name we added in the tag will be preserved. This is due to our
	// naming strategy function for columns.
	mock.ExpectExec(
		"INSERT INTO `tables` \\(`ThisIsCamelCase`, `regular_column`\\)",
	).WithArgs(
		"first custom", "first regular",
		"second custom", "second regular",
	).WillReturnResult(
		sqlmock.NewResult(1, 1),
	)

	// Insert a multi statement to test generated SQL.
	err = gormbulk.BulkInsert(gdb, []interface{}{
		Table{
			RegularColumn: "first regular",
			Custom:        "first custom",
		},
		Table{
			RegularColumn: "second regular",
			Custom:        "second custom",
		},
	}, 10)

	require.NoError(t, err)

	// Set the default naming strategy for native go.
	gorm.TheNamingStrategy.Column = defaultColumnStrategy

	// Add new mock expectations that will be executed by native gorm.
	mock.ExpectBegin()
	mock.ExpectExec(
		"INSERT INTO `tables` \\(`regular_column`,`ThisIsCamelCase`\\)",
	).WithArgs(
		"third regular", "third custom",
	).WillReturnResult(
		sqlmock.NewResult(1, 1),
	)
	mock.ExpectCommit()

	// Insert a single statement to test generated SQL.
	err = gdb.Create(&Table{
		RegularColumn: "third regular",
		Custom:        "third custom",
	}).Error

	require.NoError(t, err)
}
```
</details>

**EDIT**: And [here](https://github.com/jinzhu/gorm/blob/master/model_struct.go#L625-L630) is the code that handles this in `gorm`. So, according to this we should just quote whatever we got since DB-name should already be correct since it comes from a gorm scope.

**Comment by bombsimon** [comment]
Sorry for all the confusion, you pointed me in the exactly right direction in your first post so I'm not sure why I took this long way to find the same conclusions as you did. I guess I thought you would've created a PR if you knew what needed to be fixed. :)

Anyhow, the PR now contains the requested changes which will preserve column tags just like native gorm.

**Comment by bombsimon** [comment]
Now resolved and released in [v2.0.0](https://github.com/t-tiger/gorm-bulk-insert/releases/tag/v2.0.0).
